File: CodeTop/1.twoSum.cpp
/*
 * @lc app=leetcode.cn id=1 lang=cpp
 *
 * [1] 两数之和
 * [easy]
 * https://leetcode-cn.com/problems/two-sum/
 */

// @lc code=start
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> same;
        for(int i = 0; i < nums.size(); i++){
            int curr = target - nums[i];
            if(same.find(curr) != same.end()){
                return {i, same[curr]};
            }
            same[nums[i]] = i;
        }
        return {};
    }
};
// @lc code=end

#include<unordered_map>
#include<vector>
using namespace std;
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hashTable;
        for(int i = 0; i < nums.size(); i++){
            int deterNum = target - nums[i];
            if(hashTable.find(deterNum) != hashTable.end()){
                return {i, hashTable[deterNum]};
            }
            hashTable[nums[i]] = i;
        }
        return {};
    }
};

File: CodeTopGo/1.twoSum.go
/*
 * @lc app=leetcode.cn id=1 lang=golang
 *
 * [1] 两数之和
 */

// @lc code=start
// 暴力解法
func twoSum(nums []int, target int) []int {
	for i,_ := range nums{
		for j := i+1;j < len(nums);j++{
			if nums[i] + nums[j] == target{
				return []int {i,j}
			}
		}
	}
    return nil
}
//hash
func twoSum(nums []int, target int) []int {
	hashTable := make(map[int]int)
	for i,v := range nums {
		a := v
		b := target -a
		if va,ok := m[b]; ok{
			return []int {i,va}
		}
		hashTable[a] = i
	}
	return nil
}
// @lc code=end

